src/rust/engine/src/watch.rs

```diff
@@ -0,0 +1,159 @@
+    let (watch_sender, watch_receiver) = crossbeam_channel::unbounded();
+    let watcher = Arc::new(Mutex::new(
+      Watcher::new(watch_sender, Duration::from_millis(50))
+        .map_err(|e| format!("Failed to begin watching the filesystem: {}", e))?,
+    ));
+
+    let (thread_liveness_sender, thread_liveness_receiver) = crossbeam_channel::unbounded();
+    thread::spawn(move || {
+      logging::set_destination(logging::Destination::Pantsd);
+      loop {
+        let event_res = watch_receiver.recv_timeout(Duration::from_millis(100));
+        let graph = if let Some(g) = graph.upgrade() {
+          g
+        } else {
+          // The Graph has been dropped: we're done.
+          break;
+        };
+        match event_res {
+          Ok(Ok(ev)) => {
+            let paths: HashSet<_> = ev
+              .paths
+              .into_iter()
+              .map(|path| {
+                // relativize paths to build root.
+                let mut paths_to_invalidate: Vec<PathBuf> = vec![];
+                let path_relative_to_build_root = {
+                  if path.starts_with(&canonical_build_root) {
+                    path.strip_prefix(&canonical_build_root).unwrap().into()
+                  } else {
+                    path
+                  }
+                };
+                paths_to_invalidate.push(path_relative_to_build_root.clone());
+                if let Some(parent_dir) = path_relative_to_build_root.parent() {
+                  paths_to_invalidate.push(parent_dir.to_path_buf());
+                }
+                paths_to_invalidate
+              })
+              .flatten()
+              .collect();
+            info!("notify invalidating {:?} because of {:?}", paths, ev.kind);
+            InvalidationWatcher::invalidate(&graph, &paths, "notify");
+          }
+          Ok(Err(err)) => {
+            if let notify::ErrorKind::PathNotFound = err.kind {
+              warn!("Path(s) did not exist: {:?}", err.paths);
+              continue;
+            } else {
+              error!("File watcher failing with: {}", err);
+              break;
+            }
+          }
+          Err(RecvTimeoutError::Timeout) => continue,
+          Err(RecvTimeoutError::Disconnected) => {
+            // The Watcher is gone: we're done.
+            break;
+          }
+        };
+      }
+      debug!("Watch thread exiting.");
+      // Signal that we're exiting (which we would also do by just dropping the channel).
+      let _ = thread_liveness_sender.send(());
+    });
+
+    Ok(InvalidationWatcher {
+      watcher,
+      liveness: thread_liveness_receiver,
+    })
+  }
+
+  ///
+  /// Watch the given path non-recursively.
+  ///
+  pub fn watch(&self, path: PathBuf) -> BoxFuture<(), ()> {
+    let watcher = self.watcher.clone();
+    let path2 = path.clone();
+    futures::lazy(move || watcher.lock().watch(path, RecursiveMode::NonRecursive))
```

Comment:
I think that if you reference path here, you can avoid giving it away, which avoid the `path2` clone:
```watcher.lock().watch(&path, RecursiveMode::NonRecursive)```

(because the watch method takes `P: AsRef<Path>` for the path for flexibility)